state/relationunit.go

```diff
@@ -76,133 +76,138 @@ func (ru *RelationUnit) EnterScope(settings map[string]interface{}) error {
-	//   breaks a bunch of tests in a boring but noisy-to-fix way, and is
-	//   being saved for a followup).
 	relationDocID := ru.relation.doc.DocID
-	var ops []txn.Op
-	if ru.isLocalUnit {
-		ops = append(ops, txn.Op{
-			C:      unitsC,
-			Id:     ru.unitName,
-			Assert: isAliveDoc,
-		})
-	}
-	ops = append(ops, txn.Op{
-		C:      relationsC,
-		Id:     relationDocID,
-		Assert: isAliveDoc,
-		Update: bson.D{{"$inc", bson.D{{"unitcount", 1}}}},
-	})
-
-	// * Create the unit settings in this relation, if they do not already
-	//   exist; or completely overwrite them if they do. This must happen
-	//   before we create the scope doc, because the existence of a scope doc
-	//   is considered to be a guarantee of the existence of a settings doc.
-	settingsChanged := func() (bool, error) { return false, nil }
-	settingsColl, sCloser := db.GetCollection(settingsC)
-	defer sCloser()
-	if count, err := settingsColl.FindId(ruKey).Count(); err != nil {
-		return err
-	} else if count == 0 {
-		ops = append(ops, createSettingsOp(settingsC, ruKey, settings))
-	} else {
-		var rop txn.Op
-		rop, settingsChanged, err = replaceSettingsOp(ru.st.db(), settingsC, ruKey, settings)
-		if err != nil {
-			return err
-		}
-		ops = append(ops, rop)
-	}
-
-	// * Create the scope doc.
-	ops = append(ops, txn.Op{
-		C:      relationScopesC,
-		Id:     ruKey,
-		Assert: txn.DocMissing,
-		Insert: relationScopeDoc{
-			Key: ruKey,
-		},
-	})
 
-	// * If the unit should have a subordinate, and does not, create it.
+	var settingsChanged func() (bool, error)
 	var existingSubName string
-	if subOps, subName, err := ru.subordinateOps(); err != nil {
-		return err
-	} else {
-		existingSubName = subName
-		ops = append(ops, subOps...)
-	}
+	prefix := fmt.Sprintf("unit %q in relation %q: ", ru.unitName, ru.relation)
+
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		// Before retrying the transaction, check the following
+		// assertions:
+		if attempt > 0 {
+			if count, err := relationScopes.FindId(ruKey).Count(); err != nil {
+				return nil, errors.Trace(err)
+			} else if count != 0 {
+				// The scope document exists, so we're actually already in scope.
+				return nil, nil
+			}
 
-	// Now run the complete transaction, or figure out why we can't.
-	if err := ru.st.db().RunTransaction(ops); err != txn.ErrAborted {
-		return err
-	}
-	if count, err := relationScopes.FindId(ruKey).Count(); err != nil {
-		return err
-	} else if count != 0 {
-		// The scope document exists, so we're actually already in scope.
-		return nil
-	}
+			// The relation or unit might no longer be Alive. (Note that there is no
+			// need for additional checks if we're trying to create a subordinate
+			// unit: this could fail due to the subordinate applications not being Alive,
+			// but this case will always be caught by the check for the relation's
+			// life (because a relation cannot be Alive if its applications are not).)
+			relations, rCloser := db.GetCollection(relationsC)
```

Comment:
Might make sense to move the `.GetCollections(...)` calls to outside the buildTxn func. Not gonna block it over this, since I've seen this in other places.